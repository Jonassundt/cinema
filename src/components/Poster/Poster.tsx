import React, { useState, useContext, useEffect } from "react";

import PosterElement from "../PosterElement/PosterElement";
import FetchMovie from "./fetchData";

import posterImages from "../Images";

import Parameters from "../../Parameters";

import styles from "./poster.module.css";


function Poster() {
  const { params, setParams } = useContext(Parameters);

  let height = params.fullscreen ? 1500 : 800;

  const posters = posterImages.map((poster) => (
    <PosterElement poster={poster} height={height} />
  ));

  const setPosterIndex = (posterIndex: number) => setParams({ ...params, posterIndex });
  const [leftAnimation, setLeftAnimation] = useState(false);
  const [rightAnimation, setRightAnimation] = useState(false);
  const doneLoading = () => setParams({ ...params, loading: false })
  const [slideshowId, setSlideshowId] = useState(0);
  const [slideshowIndex, setSlideshowIndex] = useState(0);

  const prevPoster = () => {
    if (leftAnimation || rightAnimation) return;
    setLeftAnimation(true);
    setTimeout(() =>
      setPosterIndex((params.posterIndex + posters.length - 1) % posters.length), 400); //Ensures that modulo works correctly.
    setTimeout(() => setLeftAnimation(false), 800);
    sessionStorage.setItem("posterIndex", ((params.posterIndex + posters.length - 1) % posters.length).toString());
  }

  const nextPoster = (slideshowIndex: number = -1) => {
    if (leftAnimation || rightAnimation) return;
    setRightAnimation(true);
    setTimeout(() =>
      slideshowIndex >= 0 ? setPosterIndex(slideshowIndex) : setPosterIndex((params.posterIndex + 1) % posters.length), 400);
    setTimeout(() => setRightAnimation(false), 800)
    sessionStorage.setItem("posterIndex", ((params.posterIndex + 1) % posters.length).toString());
  }

  const playSound = () => {
    let number = Math.random();
    let url;
    if (number < 0.1) {
      number = number * 10;
      if (number > 0.6) {
        url = "/sounds/villager.mp3";
      }
      else if (number > 0.3) {
        url = "/sounds/o.mp3"
      }
      else {
        url = "/sounds/pig.mp3";
      }
      const audio = new Audio(url);
      audio.load();
      audio.play();
      audio.volume = 0.5;
    }
  }

  const keyboardEvent = (c: Number) => {
    if (c === 37) prevPoster();
    else if (c === 39) nextPoster();
    return;
  };

  const startSlideshow = () => {
    setSlideshowIndex((params.posterIndex + 1) % posters.length);
    const id = setInterval(() => { setSlideshowIndex(prev => (prev + 1) % posters.length); }, 15000);
    setSlideshowId(parseInt(id.toString()));
  }

  const stopSlideshow = () => {
    clearInterval(slideshowId);
    setSlideshowId(0);
    setSlideshowIndex(-1);
  }

  useEffect(() => {
    document.addEventListener("keydown", event => keyboardEvent(event.keyCode), false);

    if (params.loading) {
      setTimeout(() => doneLoading(), 4900);
    }
    else if (params.slideshow && !slideshowId) { //hvis det ikke kjøres slideshow nå, og slideshow-variabel er sann, start slideshow
      startSlideshow();
    }
    else if (!params.slideshow && slideshowId) { //hvis det kjøres slideshow nå, og slideshow-variabel er false -> stopp slideshow
      stopSlideshow();
    }
    else if (params.slideshow && slideshowIndex > -1) {
      nextPoster(slideshowIndex);
    }
  }, [params.slideshow, params.posterIndex, slideshowId, slideshowIndex, setSlideshowIndex]);

  if (params.loading) {
    return (
      <div>
        <div className={params.fullscreen ? styles.PosterFullscreen : styles.Poster} onClick={doneLoading} >
          <div className={styles.Frame} >
            <div className={styles.Loader} style={params.fullscreen ? { height: "1500px", width: "100%" } : { height: "800px", width: "1300px" }}>
              <div className={styles.Circle1}></div>
              <div className={styles.Circle2}></div>
              <div className={styles.Niddle}></div>
              <div className={styles.Number}>
                {["5", "4", "3", "2", "1"].map((number) => <div key={number}>{number}</div>)}
              </div>
            </div>
          </div>
        </div>
        {params.fullscreen ? null :
          <div className={styles.Description}>
            <div className={styles.DescriptionText}>
            </div>
          </div>}
      </div>
    )
  }

  return (
    <div>
      <div className={params.fullscreen ? styles.PosterFullscreen : styles.Poster}>
        {params.slideshow ? <div /> :
          <img
            className={styles.ArrowButtons}
            src="icons/leftarrow.svg"
            alt="left arrow"
            onClick={() => { prevPoster(); playSound(); }}
          />}
        <div className={leftAnimation ? styles.LeftAnimation : rightAnimation ? styles.RightAnimation : styles.Frame} style={{ filter: `grayscale(${params.color ? 0 : 1})` }}>
          {posters[params.posterIndex]}
        </div>
        {params.slideshow ? <div /> :
          <img
            className={styles.ArrowButtons}
            src="icons/rightarrow.svg"
            alt="right arrow"
            onClick={() => { nextPoster(); playSound() }}
          />}
      </div>
      {params.fullscreen ? null :
        <div className={styles.Description}>
          {leftAnimation || rightAnimation ? null : (
            <div className={styles.DescriptionText}>
              <FetchMovie />
            </div>
          )}
        </div>}
    </div>
  );
}

export default Poster;
