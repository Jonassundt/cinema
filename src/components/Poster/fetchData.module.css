.container {
    display: flex;
    height: auto;
    width: auto;
    flex-direction: column;
    justify-content: center;
    text-align: center;
    color: rgb(185, 185, 185);
}

.title {
    font-size: 4.5em;
    line-height: 1em;
    font-weight: 600;
    max-width: 600px;
    margin: auto;
    margin-top: -0.5em;
    margin-bottom: 0.5em;
}

.tagline {
    background-color: #5d5d5d6b;
    padding: 20px;
    border-radius: 12px;
    font-size: 1.3em;
    line-height: 1.3em;
    font-style: italic;
    margin-left: auto;
    color: rgb(214, 214, 214);
    max-width: 200px;
}

.overview {
    font-size: 1.4em;
    line-height: 1.5em;
    padding: 0 50px;
    padding-bottom: 30px;
}