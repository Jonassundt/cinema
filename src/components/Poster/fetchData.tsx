import React from "react";

import Parameters from "../../Parameters";

import styles from "./fetchData.module.css";

//bond-skyfall, inception, interstellar, batman, shawshank, starwars
const movieIds = [37724, 27205, 157336, 272, 278, 1891];

export default class fetchMovie extends React.Component {
  /* class component that handles fetching from themoviedb API.
   *
   */
  static contextType = Parameters;

  state = {
    /* Contains:
     *Loading: Bool for not displaying data while data is not yet fetched
     * movie : relevant JSON objects fetched from API, used on the site
     * apiKEY : api key needed for access to API
     * posterIndex : index for posters, deciding which movie the API should be called for information on
     */
    loading: true,
    movie: { title: "", tagline: "", overview: "" },
    movieIds,
    //Ja, du ser rett. API-nÃ¸kkelen ligger her og lurer.
    apiKey: "4d785ede4f71452009daf3f954680323",
    posterIndex: -1,
  };

  fetchData = async () => {
    /* Async method, fetching the relevant JSON objects
     */
    const posterIndex = this.context.params.posterIndex;
    const response = await fetch(
      `https://api.themoviedb.org/3/movie/${movieIds[posterIndex]}?api_key=${this.state.apiKey}`
    );
    const data = await response.json();
    this.setState({ movie: data, loading: false, posterIndex });
  };

  async componentDidMount() {
    /* LifeCycle method, fetching data when component is mounted
     */
    await this.fetchData();
  }

  async componentDidUpdate() {
    /* LifeCycle method, updating movie data if another movie is displayed in Poster.tsx
     */
    if (this.context.params.posterIndex !== this.state.posterIndex)
      await this.fetchData();
  }

  render() {
    /* Renders content, changes depending on which movie information is to be displayed
     */
    return (
      <div>
        {this.state.loading || !this.state.movie ? (
          <div>Loading...</div>
        ) : (
          <div className={styles.Container}>
            <div className={styles.Tagline}>"{this.state.movie.tagline}"</div>
            <div className={styles.Title}>{this.state.movie.title}</div>
            <div className={styles.Overview}>{this.state.movie.overview}</div>
          </div>
        )}
      </div>
    );
  }
}
