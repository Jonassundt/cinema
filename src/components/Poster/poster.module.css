.Poster,
.PosterFullscreen {
  background-color: #5d5d5d8e;
  height: auto;
  display: flex;
  justify-content: space-between;
  align-items: center;
  border-radius: 20px;
  min-width: 400px;
  overflow: hidden;
}

.PosterFullscreen {
  background-color: #5d5d5d8e;
  min-width: 90%;
  position: absolute;
  left: 45%;
  margin-left: -40%;
  border-radius: 12px;
}

.Frame,
.LeftAnimation,
.RightAnimation {
  height: 800px;
}

.PosterFullscreen .Frame,
.PosterFullscreen .LeftAnimation,
.PosterFullscreen .RightAnimation {
  height: 1500px;
}

.Icon {
  padding: 20px 40px;
  text-align: end;
  font-size: 3em;
  cursor: pointer;
}

.ArrowButtons,
.ArrowGrayed {
  width: 90px;
  filter: opacity(40%);
  margin: 0 15px;
  transition: 0.2s;
}

.ArrowGrayed {
  width: 90px;
  filter: opacity(10%);
  margin: 0 15px;
  transition: 0.2s;
}

.ArrowButtons:hover {
  width: 120px;
  filter: opacity(100%);
  margin: 0;
  cursor: pointer;
}

.Description,
.DescriptionFullscreen {
  margin-top: 20px;
  padding: 60px 100px 60px 100px;
  border-radius: 20px;
  color: rgb(189, 189, 189);
  background: #3333338e;
  line-height: 1.6em;
  min-height: 150px;
}

.DescriptionFullscreen {
  position: absolute;
  top: 1640px;
  margin-top: 140px;
  max-width: 80%;
  left: 45%;
  margin-left: -40%;
  border-radius: 12px;
}

.DescriptionText {
  animation: fadein 2s;
}

@keyframes left {
  0% {
    margin-left: 0;
    opacity: 100%;
  }
  40% {
    margin-left: -200%;
    opacity: 0;
    height: 500px;
  }
  100% {
    margin-left: 0;
    opacity: 100%;
  }
}

@keyframes right {
  0% {
    margin-right: 0;
    opacity: 100%;
  }
  40% {
    margin-right: -200%;
    opacity: 0;
    height: 500px;
  }
  100% {
    margin-right: 0;
    opacity: 100%;
  }
}

@keyframes fadein {
  0% {
    opacity: 0;
  }
  100% {
    opacity: 100%;
  }
}

.LeftAnimation {
  animation: left 1.2s ease-in-out;
}

.RightAnimation {
    animation: right;
    animation-duration: 1.2s;
}

.fetchMovie {
    display: flex;
    flex-direction: row;
    justify-content: center;
    align-items: center;   
}

.fetchMovieSubdiv {
    display: flex;
    flex-direction: column;
    justify-content: center;
    align-items: center;
}
