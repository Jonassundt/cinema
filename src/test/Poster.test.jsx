import React from 'react';
import renderer from 'react-test-renderer';
import Poster from '../components/Poster/Poster.tsx';

describe('testing the poster', () => {
    it("matches the snapshot", () => {
        const tree = renderer.create(<Poster />).toJSON();

        expect(tree).toMatchSnapshot();
    })
});