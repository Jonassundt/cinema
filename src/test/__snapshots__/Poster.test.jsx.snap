// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`testing the poster matches the snapshot 1`] = `
<div>
  <div
    className="Poster"
  >
    <img
      alt="<"
      className="ArrowButtons"
      onClick={[Function]}
      src="icons/leftarrow.svg"
    />
    <div
      className="Frame"
      style={
        Object {
          "filter": "grayscale(1)",
        }
      }
    >
      <svg
        className=" "
        height={800}
        name="Jamesbond"
        width={600.1500375093774}
      >
        jamesbond.svg
      </svg>
    </div>
    <img
      alt=">"
      className="ArrowButtons"
      onClick={[Function]}
      src="icons/rightarrow.svg"
    />
  </div>
  <div
    className="DescriptionLight"
  >
    <div
      className="DescriptionText"
    >
      <div>
        <div>
          Loading...
        </div>
      </div>
    </div>
  </div>
</div>
`;
